### PR TITLE
Fixed #24966 -- Made an empty ALLOWED_HOSTS raise ImproperlyConfigured when DEBUG=False

### PR DESCRIPTION
**Pull request by raiderrobert**
See discussion here for details: https://groups.google.com/forum/#!topic/django-developers/nIrbGjyOsNY

### CODE REVIEW

File: tests/settings_tests/tests.py [hunk 1]
```diff
@@ -511,3 +511,45 @@ def test_session_verification_deprecation_neither(self):
...
+        with self.assertRaises(ImproperlyConfigured):
+            warnings.filterwarnings('always')
+            Settings(settings_module)
+
+    def test_debug_off_configured(self):
+        settings_module = ModuleType('fake_settings_module')
+        settings_module.DEBUG = False
+        settings_module.ALLOWED_HOSTS = ['localhost']
+        try:
+            Settings(settings_module)
```
Comment (timgraham) [inline tests/settings_tests/tests.py]: I think it's better to omit the try/except and instead say `Settings(settings_module)  # should not raise an exception`. See 071801ccff970682a799ce754431a3c3ce3d6902

File: django/conf/__init__.py [hunk 2]
```diff
@@ -133,6 +133,9 @@ def __init__(self, settings_module):
             os.environ['TZ'] = self.TIME_ZONE
             time.tzset()
 
+        if not self.ALLOWED_HOSTS and not self.DEBUG:
+            raise ImproperlyConfigured("The ALLOWED_HOSTS setting must not be empty when DEBUG setting is off.")
```
Comment (timgraham) [inline django/conf/__init__.py]: when DEBUG = False.

File: tests/settings_tests/tests.py [hunk 3]
```diff
@@ -511,3 +511,45 @@ def test_session_verification_deprecation_neither(self):
...
+    """
+    Make sure Allowed_Hosts setting raises an ImproperlyConfigured during
+    appropriate conditions.
+    """
+
+    def test_debug_off_unconfigured(self):
+        settings_module = ModuleType('fake_settings_module')
+        settings_module.DEBUG = False
+        settings_module.ALLOWED_HOSTS = []
+        with self.assertRaises(ImproperlyConfigured):
```
Comment (timgraham) [inline tests/settings_tests/tests.py]: it's helpful to use assertRaisesMessage to ensure this is the exception we expect

File: tests/settings_tests/tests.py [hunk 4]
```diff
@@ -511,3 +511,45 @@ def test_session_verification_deprecation_neither(self):
             warnings.filterwarnings('always')
             Settings('fake_settings_module')
         self.assertEqual(len(warn), 0)
+
+
+class TestAllowedHostsSettings(unittest.TestCase):
+    """
+    Make sure Allowed_Hosts setting raises an ImproperlyConfigured during
```
Comment (timgraham) [inline tests/settings_tests/tests.py]: ImproperlyConfigured should be raised if DEBUG=False and ALLOWED_HOSTS is empty.

File: tests/settings_tests/tests.py [hunk 5]
```diff
@@ -511,3 +511,45 @@ def test_session_verification_deprecation_neither(self):
...
+    appropriate conditions.
+    """
+
+    def test_debug_off_unconfigured(self):
+        settings_module = ModuleType('fake_settings_module')
+        settings_module.DEBUG = False
+        settings_module.ALLOWED_HOSTS = []
+        with self.assertRaises(ImproperlyConfigured):
+            warnings.filterwarnings('always')
```
Comment (timgraham) [inline tests/settings_tests/tests.py]: I think this line is unneeded.

File: tests/settings_tests/tests.py (hunk 1 above)
Comment (raiderrobert) [inline tests/settings_tests/tests.py]: Changed! Thanks!

File: django/conf/__init__.py (hunk 2 above)
Comment (raiderrobert) [inline django/conf/__init__.py]: Done!

File: tests/settings_tests/tests.py (hunk 3 above)
Comment (raiderrobert) [inline tests/settings_tests/tests.py]: Ah, I see because it could have been something else that raised the issue.

Comment (raiderrobert) [inline tests/settings_tests/tests.py]: OK, added that in now too.

File: tests/settings_tests/tests.py (hunk 5 above)
Comment (timgraham) [inline tests/settings_tests/tests.py]: I think `warnings.filterwarnings('always')` is unneeded.

File: tests/settings_tests/tests.py (hunk 4 above)
Comment (timgraham) [inline tests/settings_tests/tests.py]: Could you update the docstring as suggested above?

File: tests/settings_tests/tests.py [hunk 6]
```diff
@@ -511,3 +511,40 @@ def test_session_verification_deprecation_neither(self):
...
+    """
+    Make sure Allowed_Hosts setting raises an ImproperlyConfigured during
+    appropriate conditions.
+    """
+
+    def test_debug_off_unconfigured(self):
+        settings_module = ModuleType('fake_settings_module')
+        settings_module.DEBUG = False
+        settings_module.ALLOWED_HOSTS = []
+        with self.assertRaisesMessage(ImproperlyConfigured,
```
Comment (timgraham) [inline tests/settings_tests/tests.py]: I think slightly nicer is:
```
msg = '...'
with self.assertRaisesMessage(ImproperlyConfigured, msg):
```

File: tests/settings_tests/tests.py [hunk 7]
```diff
@@ -511,3 +511,40 @@ def test_session_verification_deprecation_neither(self):
...
+                                      "The ALLOWED_HOSTS setting must not be empty when DEBUG = False."):
+            warnings.filterwarnings('always')
+            Settings(settings_module)
+
+    def test_debug_off_configured(self):
+        settings_module = ModuleType('fake_settings_module')
+        settings_module.DEBUG = False
+        settings_module.ALLOWED_HOSTS = ['localhost']
+
```
Comment (timgraham) [inline tests/settings_tests/tests.py]: think we can chop the blank lines in the last 3 tests

**Comment by timgraham** [comment]
Could you please squash commits after making those updates? Thanks.

File: tests/settings_tests/tests.py (hunk 7 above)
Comment (raiderrobert) [inline tests/settings_tests/tests.py]: got it

File: tests/settings_tests/tests.py (hunk 6 above)
Comment (raiderrobert) [inline tests/settings_tests/tests.py]: done!

**Comment by timgraham** [comment]
Here are instructions about squashing commits if it helps: https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/working-with-git/#rebasing-branches

**Comment by raiderrobert** [comment]
Think I got it figured out.

**Comment by timgraham** [comment]
buildbot, test this please.

**Comment by timgraham** [comment]
It looks like the test suite fails to run with this change.

**Comment by raiderrobert** [comment]
@timgraham, it appears that the issue is that neither runtest.py nor global_settings.py has the right settings that allow for it to start a server. How should I go about fixing this? I assume I shouldn't modify global_settings.py. So should I put DEBUG=True in runtest.py?

Also, in case you want to see it, here's my stack trace:

```
(django-repo)Roberts-MBP:tests rroskam$ PYTHONPATH=..:$PYTHONPATH ./runtests.py
Testing against Django installed in '/Users/rroskam/django/django'
Traceback (most recent call last):
  File "./runtests.py", line 435, in <module>
    options.debug_sql)
  File "./runtests.py", line 239, in django_tests
    state = setup(verbosity, test_labels)
  File "./runtests.py", line 108, in setup
    'INSTALLED_APPS': settings.INSTALLED_APPS,
  File "/Users/rroskam/django/django/conf/__init__.py", line 47, in __getattr__
    self._setup(name)
  File "/Users/rroskam/django/django/conf/__init__.py", line 43, in _setup
    self._wrapped = Settings(settings_module)
  File "/Users/rroskam/django/django/conf/__init__.py", line 137, in __init__
    raise ImproperlyConfigured("The ALLOWED_HOSTS setting must not be empty when DEBUG = False.")
django.core.exceptions.ImproperlyConfigured: The ALLOWED_HOSTS setting must not be empty when DEBUG = False.
```

**Comment by timgraham** [comment]
Tests are documented as running with `DEBUG=False` -- we can't change that. We could make `runtests.py` set a value for `ALLOWED_HOSTS` but that won't solve the `manage.py test` case. If we could disable the check when testing that might be ideal, but not sure about how to go about that.

**Comment by carljm** [comment]
We don't really have an "in testing mode" flag, but we do have `setup_test_environment` and `teardown_test_environment`, so I suppose we could monkeypatch something there, or add something to `ALLOWED_HOSTS`.

Actually we already do that, it just doesn't happen soon enough to help in this case.

I am wondering if making it a hard fail at settings-loading time like this is too aggressive. Maybe we should go with a deploy-check instead: https://docs.djangoproject.com/en/1.8/ref/django-admin/#django-admin-option---deploy

**Comment by timgraham** [comment]
deploy check would be fine with me.

**Comment by raiderrobert** [comment]
OK, would you recommend they go under `core/checks/security/base.py`? That's where "You should not have DEBUG set to True in deployment." and other such deployment warnings are at.

**Comment by timgraham** [comment]
It's loosely security related so that seems okay to me since we don't have another alternative right now.

**Pull request closed without merging**